ofrak_core/ofrak/core/llm.py

```diff
@@ -0,0 +1,172 @@
+        }
+
+        async with aiohttp.ClientSession() as session:
+            async with session.post(config.api_url, json=body, headers=headers) as response:
+                response.raise_for_status()
+                data = await response.json()
+                if "message" in data:
+                    message = data["message"]
+                elif "choices" in data and data["choices"]:
+                    message = data["choices"][0]["message"]
+                return LlmAttributes(message["content"])
+
+
+class LlmFunctionAnalyzer(Analyzer[LlmAnalyzerConfig, LlmAttributes]):
+    # Targets ComplexBlocks, but we don't want it to run automatically
+    targets = ()
+    outputs = (LlmAttributes,)
+
+    async def analyze(self, resource: Resource, config: CC) -> AnalyzerReturnType:
+        pass
+
+
+def indent(s: str, spaces: int = 2) -> str:
+    return "\n".join(" " * spaces + line for line in s.splitlines())
+
+
+def make_serializable(o):
+    if o is None:
+        return o
+    elif isinstance(o, (int, float)):
+        return o
+    elif isinstance(o, str):
+        return o
+    elif isinstance(o, (list, set)):
+        return [make_serializable(x) for x in o]
+    elif isinstance(o, dict):
+        return {make_serializable(k): make_serializable(v) for k, v in o.items()}
+    elif isinstance(o, type):
+        return o.__name__
+    elif isinstance(o, Enum):
+        return str(o)
+    elif isinstance(o, bytes):
+        return o.hex()
+    elif dataclasses.is_dataclass(o):
+        return make_serializable(dataclasses.asdict(o))
+    elif isinstance(o, os.stat_result):
+        return {
+            name: getattr(o, name)
+            for name in [
+                "st_mode",
+                "st_ino",
+                "st_dev",
+                "st_nlink",
+                "st_uid",
+                "st_gid",
+                "st_size",
+                "st_atime",
+                "st_mtime",
+                "st_ctime",
+            ]
+            if hasattr(o, name)
+        }
+    else:
+        return repr(o)
+
+
+def serialize(o: Optional[Union[int, float, str, List[Any], Dict[Any, Any]]]) -> str:
+    if o is None:
+        return "null"
+    elif isinstance(o, (int, float)):
+        return str(o)
+    elif isinstance(o, str):
+        return json.dumps(o)
+    elif isinstance(o, list):
+        return "\n".join("- " + serialize(x) for x in o)
+    elif isinstance(o, dict):
+        result = []
+        for k, v in o.items():
+            s = serialize(v)
+            if "\n" in s:
+                result.append(f"{k}:\n{indent(s)}")
+            else:
+                result.append(f"{k}: {s}")
+        return "\n".join(result)
+
+
+async def dump_attributes(resource: Resource) -> str:
+    model = resource.get_model()
+    # The data summary is un-informative and verbose
+    if DataSummary in model.attributes:
+        del model.attributes[DataSummary]
+    return serialize(make_serializable(model.attributes))
```

Comment:
Isn't this essentially a re-implementation of what the `PJSONSerializationService` does? Is there a reason to not use it?

Reply:
PJSON has types alongside the values. This is useful for deserializing PJSON values, but would make this code overly complicated, since we never need to deserialize the resulting text and thus would end up adding logic to drop the types anyway. 

Moreover, PJSON is nice as a wire format, but is not optimized for human-readable text "serialization." We want to deviate from its representation of some values to make the resulting text easier for the LLM to understand.

I also (personally) just don't like the PJSON serialization service.

Reply:
Are we sure that the serialized PJSON would hinder the LLM analysis? This sounds like a hypothesis and not something that was tested.

We should, whenever possible, avoid writing ~80 lines of code if it can be avoided.

Reply:
I could surely get something working with PJSON. The (admittedly untested) hypothesis is that the code to do so would be as long as this code, less clear than this code, and dependent on undocumented serialized PJSON structure, which may change in the future, necessitating changes here.

If you still think I should rewrite this to use PJSON, I will.